### PR TITLE
unit: add basic unit tests for custom conflict - swift

### PR DESCRIPTION
**Pull request by jayahariv**
* add basic unit tests for localWin, remoteWin, localDelete, remoteDelete, noDeleteButReturnNil

Ref: #2428

### CODE REVIEW

File: Swift/Tests/ReplicatorTest.swift [hunk 1]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
         XCTAssertEqual(db.count, 2)
     }
     
+    func testConflictHandlerRemoteWin() throws {
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
    func testConflictHandlerRemoteWins() throws {
```

File: Swift/Tests/ReplicatorTest.swift [hunk 2]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        
+        XCTAssertNil(resolver.winner)
+        XCTAssertEqual(db.count, 0)
+        XCTAssertNil(db.document(withID: "doc"))
+    }
+    
     #endif
 }
+
+class MyConflictResolver: ConflictResolver {
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
class TestConflictResolver: ConflictResolver {
```

File: Swift/Tests/ReplicatorTest.swift [hunk 3]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        doc1.setBoolean(true, forKey: "pass")
+        try db.saveDocument(doc1)
+        
+        let doc2 = otherDB.document(withID: "doc")!.toMutable()
+        doc2.setString("striped", forKey: "pattern")
+        try otherDB.saveDocument(doc2)
+        
+        // Pull:
+        config = self.config(target: target, type: .pull, continuous: false)
+        let resolver = MyConflictResolver()
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
        let resolver = TestConflictResolver()
```

File: Swift/Tests/ReplicatorTest.swift [hunk 4]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        try db.saveDocument(doc1)
+        
+        let doc2 = otherDB.document(withID: "doc")!.toMutable()
+        doc2.setString("striped", forKey: "pattern")
+        doc2.setBoolean(true, forKey: "pass")
+        try otherDB.saveDocument(doc2)
+        
+        // Pull:
+        config = self.config(target: target, type: .pull, continuous: false)
+        let resolver = MyConflictResolver()
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
        let resolver = TestConflictResolver()
```

File: Swift/Tests/ReplicatorTest.swift [hunk 5]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        doc1.setString("Hobbes", forKey: "name")
+        try db.saveDocument(doc1)
+        
+        let doc2 = otherDB.document(withID: "doc")!.toMutable()
+        doc2.setString("striped", forKey: "pattern")
+        try otherDB.saveDocument(doc2)
+        
+        // Pull:
+        config = self.config(target: target, type: .pull, continuous: false)
+        let resolver = MyConflictResolver()
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
        let resolver = TestConflictResolver()
```

File: Swift/Tests/ReplicatorTest.swift [hunk 6]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        // make changes to both documents from db and otherDB
+        try db.deleteDocument(db.document(withID: "doc")!)
+        
+        let doc2 = otherDB.document(withID: "doc")!.toMutable()
+        doc2.setString("striped", forKey: "pattern")
+        try otherDB.saveDocument(doc2)
+        
+        // Pull:
+        config = self.config(target: target, type: .pull, continuous: false)
+        let resolver = MyConflictResolver()
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
        let resolver = TestConflictResolver()
```

File: Swift/Tests/ReplicatorTest.swift [hunk 7]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        // make changes to both documents from db and otherDB
+        let doc1 = db.document(withID: "doc")!.toMutable()
+        doc1.setString("Hobbes", forKey: "name")
+        try db.saveDocument(doc1)
+        
+        try otherDB.deleteDocument(otherDB.document(withID: "doc")!)
+        
+        // Pull:
+        config = self.config(target: target, type: .pull, continuous: false)
+        let resolver = MyConflictResolver()
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
        let resolver = TestConflictResolver()
```

File: Swift/Tests/ReplicatorTest.swift [hunk 8]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        
+        let savedDoc = db.document(withID: "doc")!
+        
+        let exp: [String: Any] = ["species": "Tiger", "name": "Hobbes", "pass": 1]
+        XCTAssertEqual(resolver.winner!, savedDoc)
+        XCTAssertEqual(savedDoc.toDictionary().count, 3)
+        XCTAssertEqual(savedDoc.toDictionary().keys, exp.keys)
+    }
+    
+    func testConflictHandlerLocalWin() throws {
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
    func testConflictHandlerLocalWins() throws {
```

File: Swift/Tests/ReplicatorTest.swift [hunk 9]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        
+        let savedDoc = db.document(withID: "doc")!
+        
+        let exp: [String: Any] = ["species": "Tiger", "pattern": "striped", "pass": 1]
+        XCTAssertEqual(resolver.winner!, savedDoc)
+        XCTAssertEqual(savedDoc.toDictionary().count, 3)
+        XCTAssertEqual(savedDoc.toDictionary().keys, exp.keys)
+    }
+    
+    func testConflictHandlerDelete() throws {
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
    func testConflictHandlerNullDoc() throws {
```

File: Swift/Tests/ReplicatorTest.swift [hunk 10]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        let resolver = MyConflictResolver()
+        config.conflictResolver = resolver
+        run(config: config, expectedError: nil)
+        
+        XCTAssertNil(resolver.winner)
+        XCTAssertEqual(db.count, 0)
+        XCTAssertNil(db.document(withID: "doc"))
+    }
+    
+    func testConflictHandlerDeleteLocal() throws {
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
    func testConflictHandlerDeleteLocalWins() throws {
```

File: Swift/Tests/ReplicatorTest.swift [hunk 11]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        let resolver = MyConflictResolver()
+        config.conflictResolver = resolver
+        run(config: config, expectedError: nil)
+        
+        XCTAssertNil(resolver.winner)
+        XCTAssertEqual(db.count, 0)
+        XCTAssertNil(db.document(withID: "doc"))
+    }
+    
+    func testConflictHandlerDeleteRemote() throws {
```
Comment (jayahariv) [inline Swift/Tests/ReplicatorTest.swift]:
```suggestion
    func testConflictHandlerDeletedRemoteWins() throws {
```

File: Swift/Tests/ReplicatorTest.swift [hunk 12]
```diff
@@ -845,5 +845,180 @@ class ReplicatorTest: CBLTestCase {
...
+        XCTAssertEqual(db.count, 0)
+        XCTAssertNil(db.document(withID: "doc"))
+    }
+    
     #endif
 }
+
+class TestConflictResolver: ConflictResolver {
+    var winner: Document? = nil
+    func resolve(conflict: Conflict) -> Document? {
```
Comment (pasin) [inline Swift/Tests/ReplicatorTest.swift]: Can you add a short explanation about the logic / algorithm?

**Pull request merged**